Add Rainbow HUD

tmp64 opened the pull request:
Adds a number of cvars:
- `hud_rainbow`: enables rainbow HUD (default on)
- `hud_rainbow_sat`: saturation of rainbow HUD (default 100%)
- `hud_rainbow_val`: brightness of rainbow HUD (default 100%)
- `hud_rainbow_speed`: angular speed of rainbow HUD (default: 40 degrees/s)
- `hud_rainbow_xphase`: horizontal phase shift of rainbow HUD (default: 0.4 degrees/px)
- `hud_rainbow_yphase`: vertical phase shift of rainbow HUD (default: 0.7 degrees/px)

**Demonstration:**
*Note: Rainbow GameUI not included*
https://youtu.be/i_CHv75CZC4

That's what issue #90 was about, right?

YaLTeR commented on cl_dll/ammohistory.cpp:
```diff
@@ -125,19 +125,25 @@ int HistoryResource :: DrawAmmoHistory( float flTime )
 
 				// Draw the number
+				if (gHUD.IsRainbow())
+					gHUD.GetRainbowColor(xpos, r, g, b);
```
xpos - 10?

YaLTeR reviewed (commented):
lol that's actually great

tmp64 commented:
Rewrote the whole thing. Instead of modifying all HUD elements to use rainbow color, I point sprite and text rendering functions in `gEngfuncs` to custom wrappers that change the colors. This makes it easier to maintain the code. The only place that uses `GetRainbowColor` is CHudMessage since it adds blending to original colors for effects.

Also added `hud_rainbow_yphase` cvar for vertical phase shift and saturation and value clamping so nothing breaks on invalid input.

YaLTeR commented on cl_dll/rainbow.cpp:
```diff
@@ -0,0 +1,216 @@
+    m_pCvarRainbowYPhase = CVAR_CREATE("hud_rainbow_yphase", "0.7", FCVAR_ARCHIVE);
+
+    // Save engine functions
+    m_pfnSPR_Set = gEngfuncs.pfnSPR_Set;
```
Is it in any way possible for this to be called while functions are hooked? E.g. `_restart`. Because then it'll break.

tmp64 replied:
No, it's not. Engine unloads the client when shutting down. And if unloading fails, then `Initialize`, which initializes `gEngfuncs`, is always called before `Init`.
